### PR TITLE
Dynamic node removal

### LINKED ISSUE #44: Enhancement Request: Dynamic Route Removal and Update Support in Matchit
I am in the process of developing a system that enables the dynamic registration and removal of HTTP routes while the service is in operation.

While Matchit offers impressive functionality, it currently lacks the capability to remove and update routes. Is there a possibility of incorporating these features into the framework?

I understand that there are existing workarounds for this limitation, but having built-in support for these actions in the future would be greatly appreciated.

Thank you for your dedication and hard work!

### PR DESCRIPTION
**Pull request by Totodore**
## Add a `remove` fn to the radix tree

For the moment this feature consist only of : 
* Finding the node that need to be removed
* Once found, if the node has any children, only the `value` si taken back
* Otherwise the entire leaf is dropped.

Remaining things to do, are:
- [x] Testing node removal and ensuring that there is no false positive for the node research
- [ ] Removing parent nodes if they became empty / useless (maybe with backtracking ?)

@ibraheemdev I would like your feedback on the implementation as I don't master radix tree. 

Side note: The `Display` implementation is only for testing purposes and can be gated with a `CFG(test)` or removed

* Fixes #44

### CODE REVIEW

File: src/tree.rs [hunk 1]
```diff
@@ -699,6 +784,31 @@ impl<T> Default for Node<T> {
     }
 }
 
+#[cfg(test)]
+impl<T> std::fmt::Display for Node<T> {
```
Comment (ibraheemdev) [inline src/tree.rs:788]: Is this any different than the `Debug` implementation?

File: src/tree.rs [hunk 2]
```diff
@@ -732,3 +842,65 @@ const _: () = {
...
+    use super::{normalize_params, Node};
+
+    #[test]
+    fn test_normalization() {
+        let path = "/users/:azda/test/:test";
+        let (a, b) = normalize_params(path.to_string().into_bytes()).unwrap();
+        let a1 = std::str::from_utf8(&a).unwrap();
+        let b1 = std::str::from_utf8(&b[0]).unwrap();
+        let b2 = std::str::from_utf8(&b[1]).unwrap();
+        dbg!(a1, b1, b2);
```
Comment (ibraheemdev) [inline src/tree.rs:857]: Please replace this with an assertion.

File: src/tree.rs [hunk 3]
```diff
@@ -732,3 +842,65 @@ const _: () = {
...
+            "/bar/",
+            "/bar/*catchall",
+        ];
+        let mut root = Node::default();
+
+        for route in ROUTES {
+            root.insert(*route, ()).unwrap();
+        }
+
+        println!("{root}");
```
Comment (ibraheemdev) [inline src/tree.rs:900]: Can you add tests that make sure the tree stays consistent after removal? i.e try matching against related routes after a removal.

(Also avoid the debug/print statements in the tests).

File: tests/tree.rs [hunk 4]
```diff
@@ -727,6 +727,28 @@ macro_rules! insert_tests {
    )* };
 }
 
+macro_rules! remove_tests {
```
Comment (ibraheemdev) [inline tests/tree.rs:730]: Is this used?

**Review by ibraheemdev — CHANGES_REQUESTED** [review]
Thanks for implementing this. Looks good in general, just a few nits, and I would prefer more test coverage before merging.

File: src/tree.rs (hunk 1 above)
Comment (Totodore) [reply src/tree.rs:788]: It is a bit more formated that with the `Debug` impl. The output looks like this : 
```
/ --> false
|---s --> false
|---|---d --> false
|---|---omething/ --> false
|---|---|---:a --> false
|---get/ --> false
|---doc/ --> false
|---|---rust --> false
|---bar --> false
|---|---/ --> false
|---a --> false
|---foo/ --> false
```

Comment (Totodore) [reply src/tree.rs:788]: I could improve it to make it look like this:
```rust
//! 9          \                1
//! 3          ├s               None
//! 2          |├earch\         2
//! 1          |└upport\        3
//! 2          ├blog\           4
//! 1          |    └:post      None
//! 1          |         └\     5
//! 2          ├about-us\       6
//! 1          |        └team\  7
//! 1          └contact\        8
```
Or remove it.

**Comment by Totodore** [comment]
I tried to think about the "node shrinking optimisation" when we remove a node (merging nodes together).
What we could do is having a `visited` node stack and then unwind the stack and check if nodes can be merged.
My main issue is that it requires to hold multiple mutable references so the rust borrowing rules don't let me do that...

**Comment by Totodore** [comment]
Hello @ibraheemdev, do you have an idea when you will be able to review the tests I added after your feedback? Thanks! :)

**Comment by ibraheemdev** [comment]
The changes look good, thanks. I'll try to have this merged as part of 0.8, which should be finished soon.

**Comment by ibraheemdev** [comment]
@Totodore I think you need to make sure the denormalized params match before doing a removal. Right now `remove("/{x}")` will match and remove a route registered as `"/{id}"`. If you can fix that bug and the merge conflicts (the integration test setup was reworked), I think this can be merged.

**Comment by Totodore** [comment]
I have updated the `remove` fn to work with `{x}` params. I have two remaining questions:
* If an error occurs when normalizing the path, it is discarded and `None` is returned as if the value was not found. I think it is ok to do that maybe you would prefer to return a `Result<Option<T>>`?
* Because the path is normalized and that currently I don't check the original param name provided. If I insert `/foo/{bar}/test` and that I want to remove it later I can do it with any param name, e.g. `/foo/{azd}/test`. Should I check the orignal one when searching the node?

Also I added some tests but I'll try to add some more as soon as possible.

**Comment by ibraheemdev** [comment]
> If an error occurs when normalizing the path, it is discarded and None is returned as if the value was not found.

I think that's fine, an invalid route could never be in the router. Maybe add a note in the documentation that the function also returns `None` for invalid routes.

> Should I check the original one when searching the node?

Yes, I think this would be more intuitive behavior. You can just check that the parameter remapping on the final node is equal to the remapping created for the route.

**Comment by Totodore** [comment]
Done :)

**Comment by Totodore** [comment]
@ibraheemdev I'm currently working on optimizing the tree after a node removal. It consists of two actions:
* Rewind the branch and check if nodes can be removed (only one child without value). Stop at the first non-removable node.
* Rewind the branch and check if nodes can be merged together (only one, non-wild child that doesn't have a value).

This two actions are a bit more difficult to implement at might compromise the tree in case of error. Therefore do you think it is necessary to have it for this PR or to implement this later in another?

**Comment by ibraheemdev** [comment]
@Totodore thanks for continuing to work on this. It looks good for now, I'll take a closer look and try to have it merged this weekend.

**Comment by Totodore** [comment]
Hey @ibraheemdev any knews ? If you don't have the time to review this at the moment, I'll would still be interested to work on these two features:
> * Rewind the branch and check if nodes can be removed (only one child without value). Stop at the first non-removable node.
> * Rewind the branch and check if nodes can be merged together (only one, non-wild child that doesn't have a value).
> 
> These two actions are a bit more difficult to implement and might compromise the tree in case of error. Therefore do you think it is necessary to have them for this PR or it is better to implement this later in another one?

**Comment by ibraheemdev** [comment]
This looks great, thanks! Feel free to open a new PR if you are interested in working on this further, but you can just leave an issue for now describing the optimizations.

**Pull request merged**